jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -1302,23 +1260,22 @@ public InvocationType getInvocationType()
         @Override
         protected void onCompleteSuccess()
         {
-            while (true)
+            boolean close = false;
+            boolean wake = false;
+            synchronized (_channelState)
             {
-                State last = _state.get();
-                switch (last)
+                switch (_state)
                 {
                     case PENDING:
-                        if (!_state.compareAndSet(State.PENDING, State.ASYNC))
-                            continue;
+                        _state = State.ASYNC;
+                        if (_closeCallback != null)
+                            close = true;
                         break;
 
                     case UNREADY:
-                        if (!_state.compareAndSet(State.UNREADY, State.READY))
-                            continue;
-                        if (_last)
-                            closed();
-                        if (_channel.getState().onWritePossible())
-                            _channel.execute(_channel);
+                        _state = _last ? State.CLOSED : State.READY;
+                        close = true;
```

Comment:
This is the reason that the proxy tests are failing. We should not be closing unless we are the last write. For the same behaviour as before the `UNREADY` case should look something like this:
```java
case UNREADY:
    if (_last)
    {
        _state = State.CLOSED;
        releaseBuffer();
        close = true;
    }
    else
    {
        _state = State.READY;
    }

    wake = _channel.getState().onWritePossible();
    break;
```

Reply:
Ah! good catch!!!